answerking_app/models/serializers.py

```diff
@@ -25,7 +24,71 @@ def compress_white_spaces(value: str) -> str:
     return re.sub(" +", " ", value)
 
 
-class ItemSerializer(serializers.ModelSerializer):
+class CategoryDetailSerializer(serializers.ModelSerializer):
+    id = serializers.IntegerField(required=False)
+    name = serializers.CharField(
+        max_length=50,
+        allow_blank=False,
+        validators=[RegexValidator("^[a-zA-Z !]+$")],
+    )
+    description = serializers.CharField(
+        max_length=200,
+        allow_null=True,
+        allow_blank=True,
+        validators=[
+            RegexValidator(
+                "^[a-zA-Z .!,#]+$",
+            )
+        ],
+    )
+
+    def create(self, validated_data: dict) -> Category:
+        products: list[Product] = self.products_check(validated_data)
+        category: Category = Category.objects.create(
+            name=validated_data["name"],
+            description=validated_data["description"],
+        )
+        category.products.add(*products)
+        return category
+
+    def update(self, category: Category, validated_data: dict) -> Category:
+        if category.retired:
+            raise HttpErrorResponse(
+                status=status.HTTP_410_GONE,
+                detail="This category has been retired",
+            )
+        products: list[Product] = self.products_check(validated_data)
+        category.name = validated_data["name"]
+        category.description = validated_data["description"]
+        category.products.set(objs=products)
+        category.save()
+        return category
+
+    def products_check(self, validated_data: dict) -> list[Product]:
+        products: list[Product] = []
+        if "products" in validated_data:
+            products_data: list[OrderedDict] = validated_data["products"]
```

Comment:
products_id_list: list[int] = [prod.id for prod in validated_data["products"]]
for id in products_id_list:
    product: Product = get_object_or_404(
        Product, pk=id
    )
    if product.retired:
        ....

Reply:
oh ok yeah